src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs

```diff
@@ -480,9 +720,361 @@ private partial class KeyValueListNode
+
+        /// <summary>
+        /// Returns the TraceId as a 32 character hexadecimal string.  
+        /// </summary>
+        public string AsHexString
+        {
+            get
+            {
+                if (_asHexString == null)
+                    _asHexString = SpanToHexString(AsBytes);
+                return _asHexString;
+            }
+        }
+
+        /// <summary>
+        /// Returns the TraceId as a 32 character hexadecimal string.  
+        /// </summary>
+        public override string ToString() => AsHexString;
+
+        #region equality operators
+        public static bool operator ==(in ActivityTraceId traceId1, in ActivityTraceId traceId2)
+        {
+            return traceId1._id1 == traceId2._id1 && traceId1._id2 == traceId2._id2;
+        }
+        public static bool operator !=(in ActivityTraceId traceId1, in ActivityTraceId traceId2)
+        {
+            return traceId1._id1 != traceId2._id1 || traceId1._id2 != traceId2._id2;
+        }
+        public bool Equals(ActivityTraceId traceId)
+        {
+            return _id1 == traceId._id1 && _id2 == traceId._id2;
+        }
+        public override bool Equals(object obj)
+        {
+            if (!(obj is ActivityTraceId))
+                return false;
+            ActivityTraceId traceId = (ActivityTraceId)obj;
+            return _id1 == traceId._id1 && _id2 == traceId._id2;
+        }
+        public override int GetHashCode()
+        {
+            return _id1.GetHashCode() + _id2.GetHashCode();
+        }
+        #endregion 
+
+        #region private
+
+        /// <summary>
+        /// Sets the bytes in 'outBytes' to be random values.   outBytes.Length must be less than or equal to 16
+        /// </summary>
+        /// <param name="outBytes"></param>
+        internal static void SetToRandomBytes(Span<byte> outBytes)
+        {
+            Debug.Assert(outBytes.Length <= sizeof(Guid));     // Guid is 16 bytes, and so is TraceId 
+            Guid guid = Guid.NewGuid();
+            ReadOnlySpan<byte> guidBytes = new ReadOnlySpan<byte>(&guid, sizeof(Guid));
+            guidBytes.Slice(0, outBytes.Length).CopyTo(outBytes);
+        }
+
+        #region CONVERSION binary spans to hex spans, and hex spans to binary spans  
+        /* It would be nice to use generic Hex number conversion routines, but there 
+         * is nothing that is exposed publically and efficient */
+        /// <summary>
+        /// Converts each byte in 'bytes' to hex (thus two characters) and concatinates them
+        /// and returns the resulting string.  
+        /// </summary>
+        internal static string SpanToHexString(ReadOnlySpan<byte> bytes)
+        {
+            // TODO replace with ValueStringBuilder when available.  
+            StringBuilder sb = new StringBuilder(bytes.Length * 2);
+            foreach (byte b in bytes)
+                sb.Append(BinaryToHexDigit(b >> 4)).Append(BinaryToHexDigit(b));
+            return sb.ToString();
```

Comment:
I agree it would be nice if we had a public helper for this, but in the meantime using StringBuilder like this seems like overkill and inefficient.  Assuming the input `bytes` is relatively small, you could instead do:
```C#
Span<char> result = stackalloc char[bytes.Length];
int pos = 0;
foreach (byte b in bytes)
{
    result[pos++] = BinaryToHexDigit(b >> 4);
    result[pos++] = BinaryToHexDigit(b);
}
return result.ToString();
```
or if it's larger than you'd want to put on the stack, get a buffer from the array pool, use it in the same way, and then return it.  That's what ValueStringBuilder would do, anyway.

Reply:
I agree that I don't like the StringBuilder.  It took your suggestion to use a stackalloc span.